### PR TITLE
ARROW-8841: [C++] Add benchmark and unittest for encoding::PLAIN spaced

### PR DESCRIPTION
**Pull request by frankdjx**
Benchmark(1%, 10%, 50%, 90%, 99% null probabilities) and unittest for future SIMD optimizations.

Signed-off-by: Frank Du <frank.du@intel.com>

### CODE REVIEW

**Comment by pitrou** [comment]
Suggestions:
* Either `Int32` and `Int64`, or `Float` and `Double`, are sufficient; I don't think both at the same time are useful (they should use the same respective code paths)
* It may be nice to add tests with null probability 90% and 99%

File: cpp/src/parquet/encoding_benchmark.cc [hunk 1]
```diff
@@ -199,6 +200,153 @@ static void BM_PlainDecodingFloat(benchmark::State& state) {
...
+  bench->Args({/*size*/ BM_kPlainSpacedSize, /*null_percentage=*/1});
+  bench->Args({/*size*/ BM_kPlainSpacedSize, /*null_percentage=*/10});
+  bench->Args({/*size*/ BM_kPlainSpacedSize, /*null_percentage=*/50});
+}
+
+static void BM_PlainEncodingSpacedBoolean(benchmark::State& state) {
+  const auto num_values = state.range(0);
+  const double null_percent = static_cast<double>(state.range(1)) / 100.0;
+
+  const auto values = new bool[num_values];
```
Comment (fsaintjacques) [inline cpp/src/parquet/encoding_benchmark.cc:217]: Use std::vector.

Comment (jianxind) [reply cpp/src/parquet/encoding_benchmark.cc:217]: Unlike the Put which support both T* and vector inputs, current PutSpaced only has interface to T*. I has to use bool[] here, it still need convert to bool* if using std::vector<bool>. 
  virtual void Put(const T* src, int num_values) = 0;
  virtual void Put(const std::vector<T>& src, int num_values = -1);
  virtual void PutSpaced(const T* src, int num_values, const uint8_t* valid_bits,
                         int64_t valid_bits_offset) = 0;

Another alternate way is std::vector<uint8_t> instead, and reinterpret_cast the data of std::vector<uint8_t> to bool?

**Comment by jianxind** [comment]
> Suggestions:
> 
> * Either `Int32` and `Int64`, or `Float` and `Double`, are sufficient; I don't think both at the same time are useful (they should use the same respective code paths)
> * It may be nice to add tests with null probability 90% and 99%

Thanks. Yes the int32/int64 are the same path to float/double, even for SIMD, I removed it. 90% and 99% added also.

File: cpp/src/parquet/encoding_benchmark.cc (hunk 1 above)
Comment (fsaintjacques) [reply cpp/src/parquet/encoding_benchmark.cc:219]: I forgot about the bit-packing of std::vector<bool>. But yes, a std::vector<char> with `static_cast<bool*>(values.data()) should work. Or just use Int8Array like the other benchmark. It might be worth adding a quick comment about this discrepancy.

Comment (jianxind) [reply cpp/src/parquet/encoding_benchmark.cc]: Good catch to use Int8Array thus we can reuse the template for bool also. Thanks. Uploaded the change.

File: cpp/src/parquet/encoding_benchmark.cc [hunk 2]
```diff
@@ -199,6 +200,130 @@ static void BM_PlainDecodingFloat(benchmark::State& state) {
...
+
+  auto rand = ::arrow::random::RandomArrayGenerator(1923);
+  const auto array = rand.Numeric<ArrowType>(num_values, -100, 100, null_percent);
+  const auto valid_bits = array->null_bitmap_data();
+  const auto null_count = array->null_count();
+  const auto array_actual = arrow::internal::checked_pointer_cast<ArrayType>(array);
+
+  auto encoder = MakeTypedEncoder<ParquetType>(Encoding::PLAIN);
+  // Cast only happens for special BooleanType as it use UInt8 for the array data to match
+  // a bool* input to PutSpaced.
```
Comment (pitrou) [inline cpp/src/parquet/encoding_benchmark.cc:297]: Is `bool` guaranteed to be the same as `uint8_t`???

File: cpp/src/parquet/encoding_benchmark.cc [hunk 3]
```diff
@@ -199,6 +200,130 @@ static void BM_PlainDecodingFloat(benchmark::State& state) {
...
+
+template <>
+struct BM_SpacedEncodingTraits<FloatType> {
+  using ArrowType = typename EncodingTraits<FloatType>::ArrowType;
+  using ArrayType = typename arrow::TypeTraits<ArrowType>::ArrayType;
+  using CType = typename FloatType::c_type;
+};
+
+template <>
+struct BM_SpacedEncodingTraits<DoubleType> {
```
Comment (pitrou) [inline cpp/src/parquet/encoding_benchmark.cc:222]: Instead of repeating yourself here, you could make this the default `BM_SpacedEncodingTraits`.

Comment (jianxind) [reply cpp/src/parquet/encoding_benchmark.cc]: Thanks, done.

File: cpp/src/parquet/encoding_benchmark.cc (hunk 2 above)
Comment (jianxind) [reply cpp/src/parquet/encoding_benchmark.cc]: I just added a static_assert to guarantee the size math between the cast

File: cpp/src/parquet/encoding_benchmark.cc [hunk 4]
```diff
@@ -25,6 +25,7 @@
 #include "arrow/testing/util.h"
 #include "arrow/type.h"
 #include "arrow/util/byte_stream_split.h"
+#include "arrow/util/cpu_info.h"
```
Comment (pitrou) [inline cpp/src/parquet/encoding_benchmark.cc:28]: Why did you need this by the way?

Comment (jianxind) [reply cpp/src/parquet/encoding_benchmark.cc:28]: The size of test spaced block was set to L1 size. 
arrow::internal::CpuInfo::GetInstance()->CacheSize(arrow::internal::CpuInfo::L1_CACHE);

Comment (pitrou) [reply cpp/src/parquet/encoding_benchmark.cc:28]: Using a hardcoded size would be just as good (or even better, to compare between machines).

Comment (jianxind) [reply cpp/src/parquet/encoding_benchmark.cc]: Done

**Review by pitrou — APPROVED** [review]
+1. Thank you @jianxind .

**Comment by wesm** [comment]
I just stumbled on failures on MSVC in these benchmarks (see https://issues.apache.org/jira/browse/ARROW-8892). I'm fixing the immediate issues that I found but the problem of checking that benchmarks compile in CI will have to be done separately

In general, I think we should be less liberal about using `auto` to help avoid issues like this.

**Comment by jianxind** [comment]
Sorry that using too many autos here, I will notice it for further PRs. Thanks. Adding more case in CI is the fix otherwise I need to setup a MSVC context locally.
